sklearn/preprocessing/_function_transformer.py

```diff
@@ -72,18 +100,52 @@ class FunctionTransformer(BaseEstimator, TransformerMixin):
         Dictionary of additional keyword arguments to pass to inverse_func.
 
     """
-    def __init__(self, func=None, inverse_func=None, validate=True,
-                 accept_sparse=False, pass_y='deprecated', check_inverse=True,
-                 kw_args=None, inv_kw_args=None):
+    def __init__(self, func=None, inverse_func=None, validate=None,
+                 accept_sparse=False, force_all_finite=True,
+                 pass_y='deprecated', check_inverse=True, kw_args=None,
+                 inv_kw_args=None):
         self.func = func
         self.inverse_func = inverse_func
         self.validate = validate
         self.accept_sparse = accept_sparse
+        self.force_all_finite = force_all_finite
         self.pass_y = pass_y
         self.check_inverse = check_inverse
         self.kw_args = kw_args
         self.inv_kw_args = inv_kw_args
 
+    def _check_input(self, X):
+        # FIXME: Future warning to be removed in 0.22
+        if self.validate is None:
+            self._validate = True
+            warnings.warn("The default validate=True will be replaced by "
+                          "validate='array-or-frame' in 0.22.", FutureWarning)
+        else:
+            self._validate = self.validate
+
+        if ((not isinstance(self._validate, bool)) and
+                self._validate != 'array-or-frame'):
+            raise ValueError("'validate' should be a boolean or "
+                             "'array-or-frame'. Got {!r} instead."
+                             .format(self._validate))
+        if ((not isinstance(self.force_all_finite, bool)) and
+                self.force_all_finite != 'allow-nan'):
+            raise ValueError("'force_all_finite' should be a boolean "
+                             "or 'allow-nan'. Got {!r} instead."
+                             .format(self.force_all_finite))
+
+        if self._validate:
+            if hasattr(X, 'loc') and self._validate == 'array-or-frame':
+                if self.force_all_finite:
+                    _assert_all_finite(X.values, allow_nan=False
```

Comment:
wasn't there some case when X.values is not the same as np.array(X)? Also, this is possibly expensive, as we're converting to numpy array and then discard it.

Reply:
It is mainly the `.any().any()` that makes it slow. Once you have done `.isna()` you have a full frame of booleans, so getting the values then is cheap:

```
In [11]: %timeit df.isna().any().any()
1.87 ms ± 4.25 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [12]: %timeit df.isna()
411 µs ± 18.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [13]: %timeit df.isna().values.any()
431 µs ± 7.95 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Still not as fast as the ` _assert_all_finite` though, but much closer. 

> wasn't there some case when X.values is not the same as np.array(X)?

normally for a dataframe values will always be an array. For series this is indeed not always true, but, in the `_assert_all_finite` function `np.asanyarray` is still called, so this shouldn't matter.